File: src/medium/FindIndexFirstOccurrenceString.js
/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function (haystack, needle) {
    let i = 0;
    let j = 0;

    while (j < haystack.length && i < needle.length) {
        if (haystack[j] === needle[i]) {
            i ++;
        } else {
            j -= i;
            i = 0;
        }

        j++;
    }

    return i === needle.length ? j - i : -1;
};

File: src/easy/FindIndexFirstOccurrenceString.java
public class FindIndexFirstOccurrenceString {
    public int strStr(String haystack, String needle) {
        for (int i = 0; i < haystack.length(); i++) {
            if (checkOccurrence(haystack, needle, i)) {
                return i;
            }
        }
        return -1;
    }

    private boolean checkOccurrence(String haystack, String needle, int index) {
        for (int i = 0; i < needle.length(); i++) {
            if (index >= haystack.length()) {
                return false;
            }
            if (haystack.charAt(index) != needle.charAt(i)) {
                return false;
            }
            index++;
        }
        return true;
    }
}
